restore: gracefully handle lookup of dropped temp system tables

kev-cao opened the pull request:
Previously, if a lookup was attempted on a dropped system table (e.g. during cleanup after a failed restore), an error would be surfaced. This caused the restore jobs in the reverting state to be stuck in a retry loop. However, if we are attempting to do cleanup, we can simply skip any dropped descriptors, which this commit teaches restore to do.

Epic: CRDB-51394

Fixes: #148088

Release note: Restore no longer gets stuck in the reverting state after failed cleanup of dropped temporary system tables.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/148098)
<!-- Reviewable:end -->

msbutler reviewed (approved):
Nice! a couple things. Could you split this up into 2 commits: 1) for the regular restore changes, 2) for the online restore changes. They have different backport considerations.

Also, we don't backport to rc branches, except in extraordinary circumstances. Lets auto backport to 25.2 for now.

msbutler reviewed (commented):
oh, looks like you have a good test failure to look into

kev-cao committed:
restore: handle lookup of dropped system tables in classic restore

Previously, if a lookup was attempted on a dropped system table after
a failed restore, an error would be surfaced. This caused the restore
jobs in the reverting state to be stuck in a retry loop.  However, if
we are attempting to do cleanup, we can simply skip any dropped
descriptors, which this commit teaches restore to do.

Epic: CRDB-51394

Fixes: #148088

Release note: Classic restore no longer gets stuck in the reverting state
after failed cleanup of dropped temporary system tables.

wip

kev-cao committed:
restore: prevent retry loop after failed full cluster online restore

Previously, if online restore failed on a full cluster restore, it would
get stuck in a retry loop in its attempt to set all descriptors offline.
Those descriptors would include the temporary system tables, which were
dropped at the end of the link phase. This patch updates online restore
recovery to ignore these dropped descriptors instead of failing.

Epic: CRDB-51394

Fixes: #148088

Release note: Online restore no longer gets stuck in a retry loop when
reverting after a failed or canceled download job.

msbutler reviewed (approved):
nice investigation!

msbutler commented on pkg/backup/restore_online_test.go:
```diff
@@ -204,12 +204,71 @@ func TestOnlineRestoreRecovery(t *testing.T) {
+	filePathQuery := fmt.Sprintf(
+		`SELECT path, start_pretty, end_pretty FROM
+		(
+			SELECT row_number() OVER (), *
```
so, what exactly does this OVER() thing do? why is it guaranteed to grab a non system table file?

kev-cao replied:
The output of `SHOW BACKUP FILES` does order everything as you would expect, but it lists all the system tables first (`SHOW BACKUP FILES` iterates over `BackupManifest_Files` that are sorted with `FileCmp`). So if we always grab the last file, then unless there are no user-tables, it is guaranteed to be an SST file covering a user table.

`row_number() OVER ()` just numbers each row in the original output so that I can then reverse the order.

kev-cao commented:
bors r=msbutler
